### PR TITLE
Faster Android sync time

### PR DESCRIPTION
**Pull request by oehme**
This is a set of memory usage improvements aimed at making the sync time of large Android builds faster. It'll benefit pretty much every other build as well. Each commit is self-contained, I just didn't want to create 15 PRs ;) The commit messages each contain a deeper explanation for the change.

Given the number of changes I'd like to have 4 eyes on this one, @jjohannes and @bmuschko do you have some time for this?

### CODE REVIEW

**Comment by jjohannes** [comment]
I will have a look at this soon. If I manage later today. Otherwise on Monday morning.

**Comment by oehme** [comment]
@bmuschko @jjohannes Could you take a look at this one? I ironed out the last failing test.

**Comment by jjohannes** [comment]
Yes I haven't forgotten about this one. Some unexpected things came up today. I will look at this tomorrow.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskDependency.java [hunk 1]
```diff
@@ -40,23 +41,32 @@
...
 
     public DefaultTaskDependency() {
         this(null);
     }
 
     public DefaultTaskDependency(TaskResolver resolver) {
+        this(resolver, ImmutableSet.of());
+    }
+
+    public DefaultTaskDependency(TaskResolver resolver, ImmutableSet<Object> tasks) {
```
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskDependency.java]: `ImmutableSet<Object> tasks` -> the name _tasks_ can be confusing here. I think _values_ (as used elsewhere in this class) would be better.

File: subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultPluginCollection.java [hunk 2]
```diff
@@ -24,12 +24,14 @@
...
 import org.gradle.api.specs.Specs;
 
-public class DefaultPluginCollection<T extends Plugin> extends DefaultDomainObjectSet<T> implements PluginCollection<T> {
+import java.util.Collection;
 
-    public DefaultPluginCollection(Class<T> type) {
-        super(type);
+class DefaultPluginCollection<T extends Plugin> extends DefaultDomainObjectSet<T> implements PluginCollection<T> {
+
+    public DefaultPluginCollection(Class<T> type, Collection<T> plugins) {
```
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultPluginCollection.java]: If the class is now _package private_, both constructors should also be _package private_.

**Comment by oehme** [comment]
@jjohannes I did the requested changes on master.

Good point about how to keep the code this way. What I usually try is to make the code simpler as well as faster, so that no one would think about reverting it. In the cases where I did add more code, I'll add a bit more documentation about the reasons.

**Comment by oehme** [comment]
@jjohannes see https://github.com/gradle/gradle/commit/f60286b0fd331972b26a392628b79615ae2a1138

**Pull request merged**